New OCS endpoint to list text processing tasks

julien-nc opened the pull request:
refs https://github.com/nextcloud/text/issues/4613

This adds a `/tasks/app/{appId}` endpoint to get a list of tasks for the current user, for a specific app and optionally with a specific identifier.

While playing with the API, I discovered that it's impossible to schedule a task while not being authenticated (mostly because of a NotNull Db constraint on the user_id column). But all the endpoints are public. Maybe we should sort this out first.

* Do we want to allow non-authenticated users to schedule a job?
    * If so, do we want to be able to list all tasks that were scheduled publicly?
    * If not, should we require auth for all those endpoint?
* Are we fine with a user only being able to list his/her own tasks?
    * This would mean publicly scheduled tasks can be listed by any other non-authenticated user. Are we fine with that? 

TODO
* [x] add proper index(es)
* [x] allow authenticated users to list all their tasks by app (and optionally by identifier)

julien-nc committed:
start implementing ocs endpoint to get task list from user+appId+identifier

Signed-off-by: Julien Veyssier <julien-nc@posteo.net>

juliushaertl commented on lib/private/TextProcessing/Db/TaskMapper.php:
```diff
@@ -59,6 +59,25 @@ public function find(int $id): Task {
+			->where($qb->expr()->eq('app_id', $qb->createPositionalParameter($appId)))
+			->andWhere($qb->expr()->eq('user_id', $qb->createPositionalParameter($userId)));
+		if ($identifier !== null) {
+			$qb->andWhere($qb->expr()->eq('identifier', $qb->createPositionalParameter($identifier)));
```
iirc this column wasn’t indexed yes, so would be good to have a combined app, user, identifier index

julien-nc replied:
Yes indeed, it's not indexed.
The question would be which combination of columns makes the most sense. What will be the most common requested info? I can imagine that the assistant will list all tasks for a user with a specific state (for the result history or example).
One index by common use case should do it.

julien-nc committed:
allow anon text processing scheduling
add a textprocessing_tasks index
convert anotations to method attributes
refactor TP manager
add mapper methods

Signed-off-by: Julien Veyssier <julien-nc@posteo.net>

julien-nc commented:
What's been done:
* Convert annotations to Php method attributes
* Rename languageModelManager to textProcessingManager in the API controller
* Add endpoint for authenticated users to list their tasks by appId (and optionally by identifier)
* Add manager and mapper methods to:
    * get a task by user id and task id
    * get a list of task by user id and app id (and optionally with a specific identifier)
* Allow null user_id in `textprocessing_tasks` so tasks can be scheduled anonymously
* Add `user_id, app_id, identifier` index to `textprocessing_tasks`

@juliushaertl So it's only possible to list tasks by app while being authenticated. This is to prevent anonymous users to be able to list all tasks that were created by other anonymous users. Any problem with that in the Text integration?

juliushaertl commented:
> So it's only possible to list tasks by app while being authenticated. This is to prevent anonymous users to be able to list all tasks that were created by other anonymous users. Any problem with that in the Text integration? 

Fine with me as the MVP. I'd then limit the text integration to logged in users only for now.

juliushaertl commented on lib/private/TextProcessing/Db/TaskMapper.php:
```diff
@@ -59,6 +59,46 @@ public function find(int $id): Task {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select(Task::$columns)
+			->from($this->tableName)
+			->where($qb->expr()->eq('app_id', $qb->createPositionalParameter($appId)))
+			->andWhere($qb->expr()->eq('user_id', $qb->createPositionalParameter($userId)));
```
```suggestion
			->where($qb->expr()->eq('user_id', $qb->createPositionalParameter($userId)))
			->andWhere($qb->expr()->eq('app_id', $qb->createPositionalParameter($appId)));
```

Some databases are picky on the order, so maybe just switch them to match the index `['user_id', 'app_id', 'identifier']`

julien-nc committed:
change sql where order to match index order

Signed-off-by: Julien Veyssier <julien-nc@posteo.net>

marcelklehr commented on core/Controller/TextProcessingApiController.php:
```diff
@@ -157,4 +157,36 @@
+	 *
+	 * @param string $appId
+	 * @param string|null $identifier
+	 * @return DataResponse<Http::STATUS_OK, array{tasks: array{CoreTextProcessingTask}}, array{}>|DataResponse<Http::STATUS_INTERNAL_SERVER_ERROR, array{message: string}, array{}>
```
Is this fixable?

marcelklehr commented on lib/public/TextProcessing/IManager.php:
```diff
@@ -80,4 +80,23 @@ public function scheduleTask(Task $task) : void;
+	 * @throws NotFoundException If the task could not be found
+	 * @since 27.1.0
+	 */
+	public function getUserTask(int $id, ?string $userId): Task;
```
```suggestion
	public function getUserTask(int|float $id, ?string $userId): Task;
```

For 32bit support

julien-nc replied:
Should we then change it also in the getTask method? And what about the controller's getTask method?

marcelklehr replied:
See https://github.com/nextcloud/server/pull/39660

julien-nc replied:
Ok i'll wait for #39660 to be merged, then rebase and apply your suggested changes.
Even if it's not covered by the tests, the controller also needs this change IMO.

kesselb replied:
Sorry, I sent you on the wrong track. 

We use `int|float` sometimes in server for the files handling on 32bit, but not for http requests / database ids. 
I think it's fine to use int for the database ids.

marcelklehr commented on lib/private/TextProcessing/Manager.php:
```diff
@@ -192,7 +194,51 @@ public function getTask(int $id): OCPTask {
+	 * @throws RuntimeException If the query failed
+	 * @throws NotFoundException If the task could not be found
+	 */
+	public function getUserTask(int $id, ?string $userId): OCPTask {
```
```suggestion
	public function getUserTask(int|float $id, ?string $userId): OCPTask {
```

juliusknorr committed:
feat: Add delete task API

Signed-off-by: Julius Härtl <jus@bitgrid.net>

marcelklehr commented on core/Controller/TextProcessingApiController.php:
```diff
@@ -156,6 +156,36 @@ public function getTask(int $id): DataResponse {
+	 * 404: Task not found
+	 */
+	#[NoAdminRequired]
+	public function deleteTask(int $id): DataResponse {
```
Tasks are automatically cleaned up after a week

juliushaertl replied:
We use the new listing endpoint in text to show tasks for a document, so this would be used to keep that list at a sane length by either dismissing them or cleaning up once they were used.

juliushaertl replied:
See https://github.com/nextcloud/text/issues/4613#issuecomment-1665704873 for the most recent mockups

julien-nc committed:
bump server version to 28.0.0.2 to trigger new migration step

Signed-off-by: Julien Veyssier <julien-nc@posteo.net>

julien-nc commented:
Rebased on master, bumped server version to 28.0.0.2 to trigger the migration step. All good IMO.

juliusknorr reviewed (approved):
Seems good, tested and works for my use case. 👍 

Psalm is still not happy though.

marcelklehr reviewed (approved):
:sparkles: 

please make sure to fix psalm, though

julien-nc committed:
fix psalm issue

Signed-off-by: Julien Veyssier <julien-nc@posteo.net>

juliushaertl commented:
Failing drone acceptance test is unrelated, also there on master https://drone.nextcloud.com/nextcloud/server/38240/57/4

juliushaertl commented:
@julien-nc @marcelklehr Is there already documentation for the text processing to extend on?

juliushaertl commented:
Not triggering the backport yet, but I have it on my list of requirements for 27.1

marcelklehr commented:
@juliushaertl

Yes, text processing docs have already been merged, should be easy to extend for this :)

I've also created https://github.com/nextcloud/server/issues/39738 :)
